src/custom/state/swap/hooks.ts

```diff
@@ -0,0 +1,224 @@
+  parsedAmount
+}: {
+  currencyByInput?: Currency
+  inputCurrency?: Currency | null
+  outputCurrency?: Currency | null
+  parsedAmount?: CurrencyAmount
+  isExactIn: boolean
+}) => {
+  const tip = getTip()
+  const parsedFeeAmount = tryParseAmount(tip, inputCurrency || undefined)
+
+  // Shows fee amount in OUTPUT token
+  const feeForTradeExactIn = useTradeExactIn(isExactIn ? parsedFeeAmount : undefined, outputCurrency ?? undefined)
+    ?.inputAmount
+  // Shows fee amount in INPUT token
+  const feeForTradeExactOut = useTradeExactOut(outputCurrency ?? undefined, !isExactIn ? parsedFeeAmount : undefined)
+    ?.inputAmount
+
+  const bestTradeExactIn = useTradeExactIn(
+    isExactIn
+      ? feeForTradeExactIn && parsedAmount
+        ? tryValueToCurrency(
+            // PARSED_USER_AMOUNT - FEE_AMOUNT
+            JSBI.subtract(parsedAmount.raw, feeForTradeExactIn.raw).toString(),
+            currencyByInput
+          )
+        : parsedAmount
+      : undefined,
+    outputCurrency ?? undefined
+  )
+
+  const bestTradeExactOut = useTradeExactOut(
+    inputCurrency ?? undefined,
+    !isExactIn
+      ? feeForTradeExactOut && parsedAmount
+        ? tryValueToCurrency(
+            // INPUTAMOUNT
+            JSBI.add(parsedAmount.raw, feeForTradeExactOut.raw).toString(),
+            currencyByInput
+          )
+        : parsedAmount
+      : undefined
+  )
+
+  return {
+    bestTradeExactIn,
+    bestTradeExactOut
+  }
+}
+
+// from the current swap inputs, compute the best trade and return it.
+export function useDerivedSwapInfo(): {
+  currencies: { [field in Field]?: Currency }
+  currencyBalances: { [field in Field]?: CurrencyAmount }
+  parsedAmount: CurrencyAmount | undefined
+  v2Trade: Trade | undefined
+  inputError?: string
+  v1Trade: Trade | undefined
+} {
+  const { account } = useActiveWeb3React()
+
+  const toggledVersion = useToggledVersion()
+
+  const {
+    independentField,
+    typedValue,
+    [Field.INPUT]: { currencyId: inputCurrencyId },
+    [Field.OUTPUT]: { currencyId: outputCurrencyId },
+    recipient
+  } = useSwapState()
+
+  const inputCurrency = useCurrency(inputCurrencyId)
+  const outputCurrency = useCurrency(outputCurrencyId)
+  const recipientLookup = useENS(recipient ?? undefined)
+  const to: string | null = (recipient === null ? account : recipientLookup.address) ?? null
+
+  const relevantTokenBalances = useCurrencyBalances(account ?? undefined, [
+    inputCurrency ?? undefined,
+    outputCurrency ?? undefined
+  ])
+
+  const isExactIn: boolean = independentField === Field.INPUT
+
+  const currencyByInput = (isExactIn ? inputCurrency : outputCurrency) ?? undefined
+
+  const parsedAmount = tryParseAmount(typedValue, currencyByInput)
+
+  const { bestTradeExactIn, bestTradeExactOut } = useTradeWithTip({
+    isExactIn,
+    currencyByInput,
+    inputCurrency,
+    outputCurrency,
+    parsedAmount
+  })
+
+  const v2Trade = isExactIn ? bestTradeExactIn : bestTradeExactOut
```

Comment:
should probably rename to `trade` or `finalTrade` or `realisedTrade` or sth. Aka drop `V2` vernacular

Reply:
well didnt go for this as removing v2trade/v1trade requires changing all of the `Swap` components and that is not something i think we should do.

so i export `v2trade and v1trade` but always return `undefined` for `v1trade`

yes it's ugly, and yes i hate it, but, yes